### PR TITLE
Added HDFBatchEventsOutputer

### PR DESCRIPTION
**Pull request by Dr15Jones**
Added some additional improvements and bug fixes.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
@sabasehrish @knoepfel could you take a quick look?

File: HDFBatchEventsOutputer.cc [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    std::copy(offsets.begin(), offsets.end(), std::back_inserter(batchOffsets));
+
+    auto& blob = std::get<2>(event);
+    std::copy(blob.begin(), blob.end(), std::back_inserter(batchBlob));
+
+    //release memory
+    blob = std::vector<char>();
+  }
+
+  auto compressedBuffer  = pds::compressBuffer(0,0, compression_, compressionLevel_, batchBlob);
```
Comment (Dr15Jones) [inline HDFBatchEventsOutputer.cc:189]: I'm compressing twice, once here per batch

**Comment by Dr15Jones** [comment]
@sabasehrish @knoepfel ping

File: HDFBatchEventsOutputer.cc [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+      newContainer.release();
+      batchContainer = batchContainerPtr.load();
+    }
+  }
+
+  auto indexInBatch = eventIndex % (batchSize_);
+  //std::cout <<"indexInBatch "<<indexInBatch<<std::endl;
+  std::get<0>((*batchContainer)[indexInBatch]) = iEventID;
+  std::get<1>((*batchContainer)[indexInBatch]) = std::move(offsets);
+  std::get<2>((*batchContainer)[indexInBatch]) = std::move(buffer);
```
Comment (knoepfel) [inline HDFBatchEventsOutputer.cc:123]:
```suggestion
(*batchContainer)[indexInBatch] = std::make_tuple(iEventID, move(offsets), move(buffer));
```

File: HDFBatchEventsOutputer.cc [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+    auto& blob = std::get<2>(event);
+    //record the size of the blob as the final offset. Needed to decompress
+    // the event during reading
+    batchOffsets.push_back(blob.size());
+    std::copy(blob.begin(), blob.end(), std::back_inserter(batchBlob));
+
+    //release memory
+    blob = std::vector<char>();
+  }
```
Comment (knoepfel) [inline HDFBatchEventsOutputer.cc:198]: Some minor cleanups possible here.

```suggestion
  for(auto& [id, offsets, blob]: *batch) {
    if(index++ == eventsInBatch) {
      //batch was smaller than usual. Can happen at end of job
      break;
    }
    batchEventIDs.push_back(id);
    std::copy(offsets.begin(), offsets.end(), std::back_inserter(batchOffsets));
    //record the size of the blob as the final offset. Needed to decompress
    // the event during reading
    batchOffsets.push_back(blob.size());
    std::copy(blob.begin(), blob.end(), std::back_inserter(batchBlob));
 
    //release memory
    blob = {};
  }
```

File: HDFBatchEventsOutputer.cc [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+      auto offset = offsets[index++];
+      std::copy(s.blob().begin(), s.blob().end(), buffer.begin()+offset );
+    }
+    assert(buffer.size() == offsets[index]);
+  }
+
+  if(compressionChoice_ == CompressionChoice::kEvents or compressionChoice_ == CompressionChoice::kBoth) {
+    auto cBuffer  = pds::compressBuffer(0,0, compression_, compressionLevel_, buffer);
+
+    return {offsets, cBuffer};
```
Comment (knoepfel) [inline HDFBatchEventsOutputer.cc:304]: I believe an additional copy of `offsets` and `cBuffer` is occurring here (at least my tests on quick-bench.com suggest this).  Here and below I recommend:

```diff
-  return {offsets, cBuffer};
+  return {move(offsets), move(cBuffer)};
```

**Review by knoepfel — COMMENTED** [review]
Only minor comments, Chris.  See my comment on returning an `std::pair<std::vector<T>, std::vector<U>>`.

File: HDFBatchEventsOutputer.cc [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+namespace {
+  constexpr const char* const PRODUCTS_DSNAME="Products";
+  constexpr const char* const EVENTS_DSNAME="EventIDs";
+  constexpr const char* const OFFSETS_DSNAME="Offsets";
+  constexpr const char* const GNAME="Lumi";
+  constexpr const char* const RUN_ANAME="run";
+  constexpr const char* const LUMISEC_ANAME="lumisec";
+  constexpr const char* const COMPRESSION_ANAME="Compression";
+  constexpr const char* const COMPRESSION_LEVEL_ANAME="CompressionLevel";
+  constexpr const char* const COMPRESSION_CHOICE_ANAME="CompressionChoice";
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply HDFBatchEventsOutputer.cc:26]: I guess I like the trailing const just to emphasize the point.

File: HDFBatchEventsOutputer.cc (hunk 2 above)
Comment (Dr15Jones) [reply HDFBatchEventsOutputer.cc]: done

File: HDFBatchEventsOutputer.cc (hunk 3 above)
Comment (Dr15Jones) [reply HDFBatchEventsOutputer.cc]: done

File: HDFBatchEventsOutputer.cc (hunk 4 above)
Comment (Dr15Jones) [reply HDFBatchEventsOutputer.cc]: done

**Pull request merged**